File: C++/short-encoding-of-words.cpp
// Time:  O(n), n is the total sum of the lengths of words
// Space: O(t), t is the number of nodes in trie

class Solution {
public:
    int minimumLengthEncoding(vector<string>& words) {
        unordered_set<string> unique_words(words.cbegin(), words.cend());
        vector<TrieNode *> leaves;
        TrieNode trie;
        for (auto word : unique_words) {
            reverse(word.begin(), word.end());
            leaves.emplace_back(trie.Insert(word));
        }
        int result = 0;
        int i = 0;
        for (const auto& word: unique_words) {
            if (leaves[i++]->leaves.empty()) {
                result += word.length() + 1;
            }
        }
        return result;
    }

private:
    struct TrieNode {
        unordered_map<int, TrieNode *> leaves;
                
        TrieNode *Insert(const string& s) {
            auto* p = this;
            for (const auto& c : s) {
                if (!p->leaves[c - 'a']) {
                    p->leaves[c - 'a'] = new TrieNode;
                }
                p = p->leaves[c - 'a'];
            }
            return p;
        }
        
        ~TrieNode() {
            for (auto& node : leaves) {
                delete node.second;
            }
        }
    };
};

File: Python/short-encoding-of-words.py
# Time:  O(n), n is the total sum of the lengths of words
# Space: O(t), t is the number of nodes in trie

import collections
import functools


class Solution(object):
    def minimumLengthEncoding(self, words):
        """
        :type words: List[str]
        :rtype: int
        """
        words = list(set(words))
        _trie = lambda: collections.defaultdict(_trie)
        trie = _trie()

        nodes = [functools.reduce(dict.__getitem__, word[::-1], trie)
                 for word in words]

        return sum(len(word) + 1
                   for i, word in enumerate(words)
                   if len(nodes[i]) == 0)

